tokio-util/src/taskset.rs

```diff
@@ -0,0 +1,342 @@
+//! TaskSet API
+//!
+//! This module includes a TaskSet API for scoping the lifetimes of tasks and joining tasks in a
+//! manner similar to [`futures_util::stream::FuturesUnordered`]
+
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::runtime::Handle;
+use tokio::task::{JoinError, JoinHandle};
+
+/// A [`TaskSet`] is a mechanism for regulating the lifetimes of a set of asynchronous tasks.
+///
+/// # Task Lifetimes
+/// Tasks spawned on a [`TaskSet`] live at most slightly longer than the set that spawns them.
+/// The reason for this is that aborting a task is not an instantaneous operation.
+///
+/// Tasks could be running in another worker thread at the time they are cancelled, preventing
+/// them from being canceled immediately. As a result, when you abort those tasks using the join
+/// handles, Tokio marks those futures for cancellation. There are, however ways to await the
+/// cancellation of tasks, but they only work in an async context.
+///
+/// If you need to cancel a set and wait for the cancellation to complete, use [`Self::shutdown`].
+/// That function won't complete until *all* tasks have exited.
+#[derive(Debug)]
+pub struct TaskSet<T> {
+    tasks: Vec<JoinHandle<T>>,
+    handle: Handle,
+}
+
+impl<T> TaskSet<T> {
+    /// Constructs a new TaskSet.
+    ///
+    /// # Panic
+    /// Panics if invoked outside the context of a tokio runtime.
+    pub fn new() -> Self {
+        Self::new_with_handle(Handle::current())
+    }
+
+    /// Constructs a new TaskSet which will spawn tasks using the supplied handle.
+    pub fn new_with_handle(handle: Handle) -> Self {
+        Self {
+            tasks: Vec::new(),
+            handle,
+        }
+    }
+
+    /// Returns the amount of unjoined tasks in the set.
+    pub fn len(&self) -> usize {
+        self.tasks.len()
+    }
+
+    /// Returns true if there are no tasks left in the set.
+    pub fn is_empty(&self) -> bool {
+        self.tasks.is_empty()
+    }
+
+    /// Consume the set, waiting for all tasks on it to complete.
+    ///
+    /// # Output
+    /// The output of this function orders the results of tasks in the same order with which they
+    /// were spawned.
+    ///
+    /// # Cancellation Safety
+    /// When canceled, any tasks which have yet to complete will become detached, and the results
+    /// of completed tasks will be discarded.
+    ///
+    /// # Examples
+    /// ```
+    /// # use tokio_util::taskset::TaskSet;
+    /// # tokio_test::block_on(async {
+    /// let mut set = TaskSet::new();
+    ///
+    /// const NUMS: [u8; 5] = [0, 1, 2, 3, 4];
+    ///
+    /// for x in NUMS.iter() {
+    ///     set.spawn(async move { *x });
+    /// }
+    ///
+    /// let joined: Vec<_> = set.join_all().await.into_iter().map(|x| x.unwrap()).collect();
+    ///
+    /// assert_eq!(NUMS, joined.as_slice());
+    /// # });
+    /// ```
+    pub async fn join_all(mut self) -> Vec<Result<T, JoinError>> {
```

Comment:
> Though I suppose you could put `#[must_use]` on the method itself.

@Darksonn Currently, this might not work as expected on `async fn` (https://github.com/rust-lang/rust/issues/78149). A lint was added to rustc recently ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=78670a89965a98c217a9322941a9c3b3)), at least until `#[must_use]` is changed to apply to the future's output value instead of the `impl Future` returned by the function.

```
warning: `must_use` attribute on `async` functions applies to the anonymous `Future` returned by the function, not the value within
```